### PR TITLE
Analyze abstract code (beam files)

### PR DESCRIPTION
**Pull request by paulo-ferraz-oliveira**
[Fix inaka/elvis#500]

### CODE REVIEW

**Comment by paulo-ferraz-oliveira** [comment]
I started by porting https://github.com/inaka/elvis_core/pull/108 (to a tribute-named branch), which seems like a step in the right direction, but incomplete (and apparently abandoned), then:
1. I completed `elvis.config` (for the lib. itself) with the new `beam_files` section (is `dirs` and `filter` OK, here?),
2. I updated code to approach `ktn`,
3. I tried to apply all the same rules for `beam_files` that we have for `erl_files`,
4. I had a hard time (still not fixed!) with the following (of course source-based rules): `line_length`, `no_tabs`, and `no_trailing_whitespace`,
5. I noticed the extra following rules work on the "source" not the tree: `no_spaces`(not exposed), `macro_module_names` (🤷‍♂️), `operator_spaces` (🤷‍♂️), `max_module_length`, `max_function_length`.

I'd like a unified approach (same rule set for both `.erl` and `.beam`) but that is probably not expected or not feasible.
Thoughts?

**Comment by elbrujohalcon** [comment]
I'm not sure we should use the same rules for `erl_files` and `beam_files.
Some intersection, sure… but all the whitespace rules (as you properly noted) will not apply to beam files, neither will the rules that relate to stuff that's changed by the preprocessor, like records, includes and macros.
I think that it's reasonable to expect the ruleset for `beam_files` to be far shorter than the one for `erl_files`.

**Comment by paulo-ferraz-oliveira** [comment]
> I'm not sure we should use the same rules for `erl_files` and `beam_files`.
> Some intersection, sure… but all the whitespace rules (as you properly noted) will not apply to beam files, neither will the rules that relate to stuff that's changed by the preprocessor, like records, includes and macros.
> I think that it's reasonable to expect the ruleset for `beam_files` to be far shorter than the one for `erl_files`.

Sure. Can you then help me decide which ones will go to `beam_files`? Also, should this be noted somewhere, in the Wiki or something?

**Comment by elbrujohalcon** [comment]
A reasonable ruleset, I believe, would be…

```erlang
rules(beam_files) ->
    lists:map(
        fun (Rule) ->
            {elvis_style, Rule, elvis_style:default(Rule)}
        end,
        [ nesting_level
        , god_modules
        , no_if_expression
        , invalid_dynamic_call
        , used_ignored_variable
        , module_naming_convention
        , function_naming_convention
        , state_record_and_type
        , no_spec_with_records
        , dont_repeat_yourself
        , no_debug_call
        , variable_naming_convention
        , no_nested_try_catch
        , atom_naming_convention
        ]);
```

**Comment by paulo-ferraz-oliveira** [comment]
Should we also remove `atom_naming_convention` from that list?

**Comment by elbrujohalcon** [comment]
🤔  Is it not implemented by parsing the tree and looking for atoms?

**Review by elbrujohalcon — APPROVED** [review]
This looks good from my perspective. Is anything missing, @paulo-ferraz-oliveira ?

**Comment by elbrujohalcon** [comment]
We take care of documenting the ruleset and the possibility of analyzing beam files somewhere else, right?

**Comment by paulo-ferraz-oliveira** [comment]
> > We take care of documenting the ruleset and the possibility of analyzing beam files somewhere else, right?
> 
> [inaka/elvis#541](https://github.com/inaka/elvis/issues/541)

It's tricky to have up-to-date documentation (in `master`) while expecting release of certain stuff. One way to achieve it would be to have the same Git flow in the Wiki, but that's just not wanted, most probably.

**Comment by paulo-ferraz-oliveira** [comment]
This was a much bigger one than I anticipated.

In any case, in regards to the [previous PR](https://github.com/inaka/elvis_core/pull/137), I have to following to add, that we should consider... 

The following functions aren't exactly "beam-oriented" in analysis, which leads me to think they're be better off in `elvis_text_style`:
* `operator_spaces`
* `max_module_length`
* `max_function_length`
(though they seem to be parseable by AST - thus subject to `ignore` with MFA).

What say you?

**Comment by paulo-ferraz-oliveira** [comment]
> This looks good from my perspective. Is anything missing, @paulo-ferraz-oliveira ?

Tests were missing... loads of them 😄

**Comment by elbrujohalcon** [comment]
Oh, right… 🙈

**Comment by elbrujohalcon** [comment]
> The following functions aren't exactly "beam-oriented" in analysis, which leads me to think they're be better off in `elvis_text_style`:
> 
> * `operator_spaces`
> * `max_module_length`
> * `max_function_length`
> 
> What say you?

100% `elvis_text_style` for all of them.

**Comment by paulo-ferraz-oliveira** [comment]
OK, I'll wait until the other PR gets merged, then, to rebase and prepare this one, so that we can then move stuff to `elvis_text_style`.

**Comment by paulo-ferraz-oliveira** [comment]
Converting this to draft since I haven't had time to work on it, and it's still incomplete. (it was working up until the rebase 😢)

**Comment by paulo-ferraz-oliveira** [comment]
I've rebased on top of master, fixed all issues and have tests and static analysis to 100%.

File: config/test.config [hunk 1]
```diff
@@ -12,15 +12,24 @@
                  ],
         ruleset => erl_files
        },
-      #{dirs => ["../../"],
+      #{dirs => ["../../_build/test/lib/elvis_core/test/examples"],
+        filter => "**.beam",
```
Comment (elbrujohalcon) [inline config/test.config:16]: Why `**`?

**Review by elbrujohalcon — APPROVED** [review]

File: config/test.config (hunk 1 above)
Comment (paulo-ferraz-oliveira) [reply config/test.config:16]: Lucky Luke fingers, most likely.

Comment (paulo-ferraz-oliveira) [reply config/test.config:16]: Ah, no, because of copy-paste :-) above you'll find `**.erl`. I'll fix all occurrences.

Comment (paulo-ferraz-oliveira) [reply config/test.config]: https://github.com/inaka/elvis_core/pull/138/commits/b48eb822f9411c79f3a86ced2db1ff910626497b

**Comment by paulo-ferraz-oliveira** [comment]
Test suite's taking forever. Any way we can restart it?

**Comment by elbrujohalcon** [comment]
> Test suite's taking forever. Any way we can restart it?

Pushing another micro-change?

**Comment by paulo-ferraz-oliveira** [comment]
If you go to Travis-CI.org (or .com) - you should be able to restart this without need for a change.

**Comment by paulo-ferraz-oliveira** [comment]
It seems to be failing. I'll look at it later.

**Comment by paulo-ferraz-oliveira** [comment]
@elbrujohalcon: would it be preferable to separate `style_SUITE` from `text_style_SUITE`? For consistency it makes sense, but conceptually, both are "style" elements, so...

**Comment by paulo-ferraz-oliveira** [comment]
I still need to fix some of the tests throwing on older OTP versions (for the `beam`-specific analysis).

**Comment by paulo-ferraz-oliveira** [comment]
Most recent stuff:
* https://github.com/inaka/elvis_core/pull/138/commits/b8e271a78d7e2ef24482307efd507f9f65163b77
* https://github.com/inaka/elvis_core/pull/138/commits/a8cc849bbcd3030337fae42c7aca31360dfc2475
* https://github.com/inaka/elvis_core/pull/138/commits/f7570b903c0abc7ddb8e3e1c62c5692d4569ffe6

**Comment by paulo-ferraz-oliveira** [comment]
I think the errors are there because we're using a different `rebar3` version (and the previous one had recursive by default, already). Let me see what I can do.

**Comment by elbrujohalcon** [comment]
Tired of Travis. I just merged this.

**Comment by paulo-ferraz-oliveira** [comment]
The last check passed in all OTP versions except for 23.1, since it doesn't exist in Travis CI... so that's a win...

**Pull request merged**